### PR TITLE
python unit tests: adds new tests for graph related functions

### PR DESCRIPTION
**Pull request by PetholzA**
this commit is mainly aimed to increase testing coverage, it tests functions from dynamics.pyx and mostly graph.pyx

### CODE REVIEW

**Comment by fabratu** [comment]
@PetholzA Please rebase on top of master. So far, the PR has increased the coverage from 60% -> 66% for the graph module. If possible, it would be good to increase the coverage even more. Maybe that is done after the rebase though.

File: networkit/test/test_graph.py [hunk 1]
```diff
@@ -90,6 +90,43 @@ def testRemoveMultiEdges(self):
...
+		up1 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.NODE_ADDITION, 0, 0, 0)	
+		up2 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.NODE_ADDITION, 1, 1, 1)	
+		up3 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.NODE_ADDITION, 2, 2, 2)	
+		self.assertIsInstance(up1.__repr__(),str)
+		self.assertFalse(up1.__eq__(up2))
+		self.assertTrue(up1.__ne__(up2))
+		self.assertTrue(up1.__lt__(up2))
+		self.assertFalse(up1.__gt__(up2))
+		self.assertTrue(up1.__le__(up2))
+		self.assertFalse(up1.__ge__(up2))
```
Comment (angriman) [inline networkit/test/test_graph.py:103]: Using `__eq__`, `__ne__`, etc. should not be necessary, they enable using `==`, `<=`, etc. Please use `assertEqual`, `assertLessEqual`, etc.

Comment (angriman) [reply networkit/test/test_graph.py:103]: Also, these tests are not related with the GraphEventStream. Please move them to another test.

File: networkit/test/test_graph.py [hunk 2]
```diff
@@ -90,6 +90,43 @@ def testRemoveMultiEdges(self):
...
+		self.assertFalse(up1.__gt__(up2))
+		self.assertTrue(up1.__le__(up2))
+		self.assertFalse(up1.__ge__(up2))		
+		up4 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.NODE_ADDITION, 3, 3, 3)	
+		upE1 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.EDGE_ADDITION, 1, 2, 2.0)	
+		upE2 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.EDGE_ADDITION, 2, 3, 3.0)	
+		upE3 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.EDGE_ADDITION, 1, 3, 1.0)	
+		allUpdates=[up1,up2,up3,up4,upE1,upE2,upE3]
+		G = nk.dynamics.graphFromStream(allUpdates, weighted=True,directed=True)
+		self.assertEqual(G.numberOfNodes(), 4)
```
Comment (angriman) [inline networkit/test/test_graph.py:110]: Please also test that the edges are as expected.

File: networkit/test/test_graph.py [hunk 3]
```diff
@@ -254,6 +292,41 @@ def testAddEdgeWithMissingNodes(self):
 		self.assertEqual(G.numberOfNodes(), 3)
 		self.assertEqual(G.numberOfEdges(), 2)
 
+	def testRestoreNode(self):
+		G=nk.Graph(2)
+		G.removeNode(1)
+		G.restoreNode(1)
+		self.assertEqual(G.numberOfNodes(), 2)
```
Comment (angriman) [inline networkit/test/test_graph.py:299]:
```suggestion
		self.assertEqual(G.numberOfNodes(), 2)
		setf.assertTrue(G.hasNode(1))
```

File: networkit/test/test_graph.py [hunk 4]
```diff
@@ -254,6 +292,41 @@ def testAddEdgeWithMissingNodes(self):
...
+
+	def testEdgeWeight(self):
```
Comment (angriman) [inline networkit/test/test_graph.py:301]: The name is too generic, and you are testing different behaviors here. Please split this test in more minimal and targeted tests. For example, one test to test `increaseWeight`, another one to test `swapEdge`, another one for `totalEdgeWeight`, etc.

File: networkit/test/test_graph.py [hunk 5]
```diff
@@ -254,6 +292,41 @@ def testAddEdgeWithMissingNodes(self):
...
+			return neighborsWeights
+
+		def getInNeighborsWeight(u):
+			neighborsWeights = 0
+			for v in G.iterInNeighborsWeights(u):
+				neighborsWeights += v[1]
+			return neighborsWeights
+		
+		self.assertEqual(getNeighborsWeight(0), 3.0)
+		self.assertEqual(getInNeighborsWeight(0), 0.0)
```
Comment (angriman) [inline networkit/test/test_graph.py:328]: Here you can use `weightedDegree` and `weightedDegreeIn`.

**Review by angriman — CHANGES_REQUESTED** [review]

File: networkit/test/test_graph.py [hunk 6]
```diff
@@ -90,6 +92,47 @@ def testRemoveMultiEdges(self):
 					self.assertFalse(G.hasEdge(u, v))
 				self.assertEqual(G.numberOfEdges(), 0)
 
+	def testGraphEventOperators(self):
+		up1 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.NODE_ADDITION, 0, 0, 0)	
+		up2 = nk.dynamics.GraphEvent(nk.dynamics.GraphEventType.NODE_ADDITION, 1, 1, 1)
+		self.assertIsInstance(up1.__repr__(),str)
```
Comment (angriman) [inline networkit/test/test_graph.py:98]: I think that this is supposed to be used with `repr(up1)`. Also, we should test that the string representation works as intended, see the [implementation](https://github.com/networkit/networkit/blob/master/networkit/cpp/dynamics/GraphEvent.cpp#L17).

File: networkit/test/test_graph.py [hunk 7]
```diff
@@ -55,16 +54,17 @@ def testEdgeIterator(self):
...
-		self.assertEqual(G.numberOfNodes(), G2.numberOfNodes())
-		self.assertEqual(G.numberOfEdges(), G2.numberOfEdges())
-		self.assertEqual(G.edgeId(0,1), G2.edgeId(0,1))
-	
+		G.swapEdge(0,1, 2,3) # now edges are:(0,3, 1,0) and (2,1, 2.0)
+		self.assertTrue(G.hasEdge(0,3))
+		self.assertTrue(G.hasEdge(2,1))
+		#testing if the weight remained
+		self.assertEqual(G.totalEdgeWeight(), 3.0)					
+
```
Comment (angriman) [inline networkit/test/test_graph.py:67]: Drop the additional white spaces.

File: networkit/test/test_graph.py [hunk 8]
```diff
@@ -55,16 +54,17 @@ def testEdgeIterator(self):
...
-		pickledGraph = pickle.dumps(G)
-		G2 = pickle.loads(pickledGraph)
-		self.assertEqual(G.numberOfNodes(), G2.numberOfNodes())
-		self.assertEqual(G.numberOfEdges(), G2.numberOfEdges())
-		self.assertEqual(G.edgeId(0,1), G2.edgeId(0,1))
-	
+		G.swapEdge(0,1, 2,3) # now edges are:(0,3, 1,0) and (2,1, 2.0)
+		self.assertTrue(G.hasEdge(0,3))
+		self.assertTrue(G.hasEdge(2,1))
+		#testing if the weight remained
```
Comment (angriman) [inline networkit/test/test_graph.py:65]:
```suggestion
		# Test that the total edge weight remained the same.
```

File: networkit/test/test_graph.py [hunk 9]
```diff
@@ -119,7 +119,115 @@ def nodeInIter(node):
 			forEdgesNbrs = []
 			G.forInEdgesOf(node, nbrFunc)
 			nodeInNbrs = nodeInIter(node)
-			self.assertEqual(sorted(forEdgesNbrs), sorted(nodeInNbrs))				
+			self.assertEqual(sorted(forEdgesNbrs), sorted(nodeInNbrs))
```
Comment (angriman) [inline networkit/test/test_graph.py:122]: I guess that this change was added by mistake.

File: networkit/test/test_graph.py [hunk 10]
```diff
@@ -119,7 +119,115 @@ def nodeInIter(node):
...
+		g.removeNode(nk.graphtools.randomNode(g))
+		g.removeNode(nk.graphtools.randomNode(g))
+		doTest(g)	
+	
+	def testNodesInRandomOrder(self):
+		G = nk.Graph(3, directed=True)
+		G.addEdge(0,1)
+		G.addEdge(0,2)
+		G.addEdge(1,2)
+		G.forNodesInRandomOrder(lambda u: self.assertFalse(G.hasEdge(u, 0)))
```
Comment (angriman) [inline networkit/test/test_graph.py:217]: I don't follow the logic of this test. Shouldn't we check that the node `u` exists and that the iterator goes over all nodes?

**Review by angriman — COMMENTED** [review]
Thank you, I left a few comments. I see that there are lots of unrelated changes (moving around existing code, adding/removing empty lines or white spaces). This makes reviewing PRs harder than necessary. If that's not too difficult, could you please drop these changes and only include the actual changes that you made?

**Comment by fabratu** [comment]
The coverage is still below 80%. However, the main block missing are the node attributes. I would suggest leaving these out of this PR. Otherwise the PR might get too long.

**Comment by PetholzA** [comment]
@angriman, your requests should mostly be adressed. I still left some "redundant" tests, because the goal was, to increase python testing coverage as much as possible.

File: networkit/test/test_graph.py [hunk 11]
```diff
@@ -290,6 +290,108 @@ def getSmallGraph(self, weighted=False, directed=False):
...
+		self.assertEqual(GraphDiff.getNumberOfNodeAdditions(), 0)
+		self.assertEqual(GraphDiff.getNumberOfNodeRestorations(), 0)
+		self.assertEqual(GraphDiff.getNumberOfNodeRemovals(), 0)
+		self.assertEqual(GraphDiff.getNumberOfEdgeAdditions(), 2)					
+		self.assertEqual(GraphDiff.getNumberOfEdgeRemovals(), 0)
+		self.assertEqual(GraphDiff.getNumberOfEdgeWeightUpdates(), 0)
+
+	def testNodesInRandomOrder(self):
+		G = nk.Graph(5)
+		G.forNodesInRandomOrder(lambda u: self.assertIn(u, [0,1,2,3,4]))
```
Comment (angriman) [inline networkit/test/test_graph.py:366]:
```suggestion
		G.forNodesInRandomOrder(lambda u: self.assertTrue(G.hasNode(u)))
```

**Review by angriman — COMMENTED** [review]
Thank you, just one minor thing. The rest LGTM.

**Review by angriman — APPROVED** [review]

**Pull request merged**